File: dsa_taijutsu/queues/customQ.cpp
#include <iostream>
#include <stdexcept>
using namespace std;

template <typename T>
class Queue {
private:
    T* arr;
    int front;
    int rear;
    int capacity;
    int size;

public:
    Queue(int capacity = 10) : capacity(capacity), size(0), front(0), rear(-1) {
        arr = new T[capacity];
    }

    ~Queue() {
        delete[] arr;
    }

    void enqueue(T data) {
        if (size == capacity) {
            throw overflow_error("Queue is full");
        }
        rear = (rear + 1) % capacity;
        arr[rear] = data;
        size++;
    }

    T dequeue() {
        if (isEmpty()) {
            throw underflow_error("Queue is empty");
        }
        T data = arr[front];
        front = (front + 1) % capacity;
        size--;
        return data;
    }

    bool isEmpty() const {
        return size == 0;
    }

    int getSize() const {
        return size;
    }

    T peek() const {
        if (isEmpty()) {
            throw underflow_error("Queue is empty");
        }
        return arr[front];
    }
};

#include <iostream>
#include <stdexcept>
using namespace std;

template <typename T>
class Dequeue {
private:
    T* arr;
    int front;
    int rear;
    int capacity;
    int size;

public:
    Dequeue(int capacity = 10) : capacity(capacity), size(0), front(0), rear(-1) {
        arr = new T[capacity];
    }

    ~Dequeue() {
        delete[] arr;
    }

    void addFront(T data) {
        if (size == capacity) {
            throw overflow_error("Dequeue is full");
        }
        front = (front - 1 + capacity) % capacity;
        arr[front] = data;
        size++;
    }

    void addRear(T data) {
        if (size == capacity) {
            throw overflow_error("Dequeue is full");
        }
        rear = (rear + 1) % capacity;
        arr[rear] = data;
        size++;
    }

    T removeFront() {
        if (isEmpty()) {
            throw underflow_error("Dequeue is empty");
        }
        T data = arr[front];
        front = (front + 1) % capacity;
        size--;
        return data;
    }

    T removeRear() {
        if (isEmpty()) {
            throw underflow_error("Dequeue is empty");
        }
        T data = arr[rear];
        rear = (rear - 1 + capacity) % capacity;
        size--;
        return data;
    }

    bool isEmpty() const {
        return size == 0;
    }

    int getSize() const {
        return size;
    }
};

int main() {
    Dequeue<int> dq(5);
    dq.addRear(10);
    dq.addRear(20);
    dq.addFront(5);

    cout << "Remove front: " << dq.removeFront() << endl;
    cout << "Remove rear: " << dq.removeRear() << endl;

    return 0;
}

File: python_genjutsu/stacksQ'sHash/customQ.py
class Queue:
    def __init__(self, capacity=10):
        self.capacity = capacity
        self.queue = []
    
    def enqueue(self, data):
        if len(self.queue) >= self.capacity:
            raise OverflowError("Queue is full")
        self.queue.append(data)

    def dequeue(self):
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.queue.pop(0)
    
    def is_empty(self):
        return len(self.queue) == 0
    
    def size(self):
        return len(self.queue)

    def peek(self):
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.queue[0]

# Example Usage
q = Queue(5)
q.enqueue(10)
q.enqueue(20)
q.enqueue(30)

print("Dequeued:", q.dequeue())
print("Front:", q.peek())
print("Queue size:", q.size())


class Dequeue:
    def __init__(self, capacity=10):
        self.capacity = capacity
        self.dequeue = []
    
    def add_front(self, data):
        if len(self.dequeue) >= self.capacity:
            raise OverflowError("Dequeue is full")
        self.dequeue.insert(0, data)

    def add_rear(self, data):
        if len(self.dequeue) >= self.capacity:
            raise OverflowError("Dequeue is full")
        self.dequeue.append(data)

    def remove_front(self):
        if self.is_empty():
            raise IndexError("Dequeue is empty")
        return self.dequeue.pop(0)

    def remove_rear(self):
        if self.is_empty():
            raise IndexError("Dequeue is empty")
        return self.dequeue.pop()
    
    def is_empty(self):
        return len(self.dequeue) == 0

    def size(self):
        return len(self.dequeue)

# Example Usage
dq = Dequeue(5)
dq.add_rear(10)
dq.add_rear(20)
dq.add_front(5)

print("Removed front:", dq.remove_front())
print("Removed rear:", dq.remove_rear())
